tests/common.rs

```diff
@@ -0,0 +1,197 @@
+use std::env;
+use std::collections::HashMap;
+
+use async_trait::async_trait;
+use anyhow::{Error, Result};
+use port_selector::random_free_tcp_port;
+use tokio::process::{Child, Command};
+use bollard::Docker;
+use bollard::container::{
+    Config as ContainerConfig,
+    CreateContainerOptions,
+    StartContainerOptions,
+};
+use random_string::{generate as generate_random_string};
+
+/// Variable that determines/differentiates CI environments
+const ENV_VAR_CI: &str = "CI";
+
+/// ENV Variable that determines whether docker is enabled
+const ENV_VAR_DOCKER_DISABLED: &str = "TEST_DOCKER_DISABLED";
+const DEFAULT_DOCKER_DISABLED: bool = true;
+
+/// ENV Variable that determines where docker is located
+const ENV_VAR_DOCKER_HOST: &str = "DOCKER_HOST";
+
+/// ENV Variable that determines where the NATS is located
+const ENV_VAR_NATS_BIN_PATH: &str = "TEST_NATS_BIN";
+const DEFAULT_NATS_BIN_PATH: &str = "/usr/bin/nats-server";
+
+const ENV_VAR_NATS_DOCKER_IMAGE: &str = "TEST_NATS_DOCKER_IMAGE";
+const DEFAULT_NATS_DOCKER_IMAGE: &str = "nats:2.9.15-scratch@sha256:1a82b1b261d169994fc6162f2f25894a49e20971560f8b2dff4f90eefb32b421";
+
+/// A test instance of a given requirement (ex. NATS)
+#[async_trait]
+pub trait EphemeralResource {
+    /// Stop the instance
+    async fn stop(&mut self) -> Result<()>;
+}
+
+/// A NATS instance that can be used for test
+#[async_trait]
+pub trait NatsInstance: EphemeralResource {
+    /// Get the address of the nats instance
+    async fn get_addr(&self) -> Result<String>;
+}
+
+const NUM_PORT_TRIES: u8 = 5;
+
+/// Helper function that picks a random port
+fn pick_random_port() -> Result<u16> {
+    let mut tries = 0;
+    let mut port = None;
+    loop {
+        tries += 1;
+        if let Some(p) = random_free_tcp_port() {
+            port = Some(p);
+            break;
+        }
+        if tries >= 5 {
+            return Err(Error::msg(format!("Failed to find port after trying {NUM_PORT_TRIES} times")));
+        }
+    }
+
+    match port {
+        Some(port) => Ok(port),
+        None => Err(Error::msg("Failed to find free TCP port")),
+    }
+}
+
+/// Start a NATS instance for use during test
+pub(crate) async fn make_nats_instance() -> Result<Box<dyn NatsInstance>> {
+    let in_ci = env::var(ENV_VAR_CI).is_ok();
+    let docker_disabled = env::var(ENV_VAR_DOCKER_DISABLED).map(|v| v == "true").unwrap_or(DEFAULT_DOCKER_DISABLED);
+
+    let port = pick_random_port()?;
+
+    if docker_disabled || in_ci {
+        Ok(Box::new(ProcessNats::new(port).await?))
+    } else {
+        Ok(Box::new(DockerizedNats::new(port).await?))
+    }
+}
+
+/// A process-based NATS instance
+struct ProcessNats {
+    child: Child,
+    port: u16,
+}
+
+impl ProcessNats {
```

Comment:
In other tests that require NATS, we've just ensured or assumed that NATS was running. For example, in the wasmCloud OTP host, the CI scripts start nats as a service and the scripts to run individual tests.

It feels like a lot of extra complexity (and potentially brittle execution inside the github CI) to try and manually launch nats within the test code when it should be considerably easier to just let the environment take care of starting NATS.

Thoughts on this @thomastaylor312 ?

Reply:
Yeah I already put the stuff that starts NATS for you if it isn't already running in the `Makefile`. Would prefer to keep that part there